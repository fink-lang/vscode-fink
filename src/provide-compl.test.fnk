{describe, it, expect, mock, mock_fn} = import '@fink/jest'
{to_match_snapshot} = import '@fink/jest'


mock
  'vscode'
  fn: dict:
    CompletionItem: mock_fn fn ...completion: {completion}
    CompletionItemKind: dict:
      Variable: 'test-var'
      Function: 'test-func'

    SnippetString: mock_fn fn ...snippet: {snippet}
    MarkdownString: mock_fn fn ...markdown: dict:
      markdown
      appendCodeblock: mock_fn fn ...args:
        markdown.push args

    languages: dict:
      registerCompletionItemProvider: mock_fn()

  {virtual: true}


{languages} = import 'vscode'
{parse} = import '@fink/larix'

{add_completion_provider} = import './provide-compl'


async_parse = fn doc:
  await parse doc.getText(), `${doc.uri}`


describe 'completion provider', fn:

  it 'provides completions', fn:
    ctx = dict:
      subscriptions: dict:
        push: mock_fn()

    add_completion_provider ctx, async_parse
    {mock: {calls: [[, provider]]}} = languages.registerCompletionItemProvider

    document = dict:
      getText: fn: `
        foo = 1234
        bar = fn spam:
          fn ni:
            ni
          shr
        na = 567
      `
      uri: {test: 'test-uri:test.fnk'}
      languageId: 'fink'

    pos = {line: 4, character: 5}

    completions = await provider.provideCompletionItems document, pos

    expect
      completions
      to_match_snapshot
