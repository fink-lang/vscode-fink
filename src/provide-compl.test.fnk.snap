// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`completion provider provides completions 1`] = `
Array [
  Object {
    "completion": Array [
      "foo",
      "test-var",
    ],
  },
  Object {
    "completion": Array [
      "bar",
      "test-var",
    ],
  },
  Object {
    "completion": Array [
      "spam",
      "test-var",
    ],
  },
  Object {
    "completion": Array [
      "match",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Return the first result of \`test: result\` where foo matches \`test\`.",
        Array [
          "match foo:
  test: result
  {bar: 'spam'}: shrub
  [bar, 'spam']: shrub
  'spam': shrub
  else: ni
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "match $1:
  $2: $3
  else: $4
$0",
      ],
    },
  },
  Object {
    "completion": Array [
      "fold",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Return a function \`fn iterable:\` that reduces all items of \`iterable\`
to a single value.",
        Array [
          "pipe [1, 2, 3]:
  fold item, acc=0:
    item + acc

# == 6
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "fold $1, $2:
  $3
$0",
      ],
    },
  },
  Object {
    "completion": Array [
      "unfold",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Return a function \`fn curr:\` that generates items from the \`curr\`
value.",
        Array [
          "count = unfold curr=0:
  (curr + inc, curr + inc)

[a, b, c] = count()
# [0, 1, 2]
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "unfold $1:
  ($2, $3)
$0",
      ],
    },
  },
  Object {
    "completion": Array [
      "map",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Return a function \`fn iterable:\` that maps each \`item\` of \`iterable\`.",
        Array [
          "pipe [1, 2, 3]:
  map item:
    item * 2

# == [2, 4, 6]
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "map $1:
  $2
$0",
      ],
    },
  },
  Object {
    "completion": Array [
      "filter",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Return a function \`fn iterable:\` that only yields \`item\` of \`iterable\`
for which the block returns \`true\`.",
        Array [
          "pipe [1, 2, 3, 4]:
  filter item:
    item % 2

# == [2, 4]
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "filter $1:
  $2
$0",
      ],
    },
  },
  Object {
    "completion": Array [
      "while",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Return a function \`fn iterable:\` that yields each \`item\` of \`iterable\`
while the block returns true",
        Array [
          "pipe [1, 2, 3, 4]:
  while item:
    item < 3

# == [1, 2]
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "while $1:
  $2
$0",
      ],
    },
  },
  Object {
    "completion": Array [
      "find",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Return a function \`fn iterable:\` that returns the first \`item\` of
\`iterable\` for which the block returns true",
        Array [
          "pipe [1, 2, 3, 4]:
  find item:
    item > 2

# == 3
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "find $1:
  $2
$0",
      ],
    },
  },
  Object {
    "completion": Array [
      "pipe",
      "test-func",
    ],
    "documentation": Object {
      "appendCodeblock": [MockFunction],
      "markdown": Array [
        "Calls all expressions with the result of the previous call, starting
with the pipe arg for the first call.",
        Array [
          "pipe [1, 2, 3, 4]:
  map item:
    item * 2
  find item:
    item > 4

# == 6
",
          "fink",
        ],
      ],
    },
    "insertText": Object {
      "snippet": Array [
        "pipe $1:
  $2
$0",
      ],
    },
  },
]
`;
